From b3c8e4f39dc35b89401a3da4c5f018a5886595d7 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:39:35 -0400
Subject: [virt] kvm/x86: Fix page-tables reserved bits

Message-id: <1406040016-3289-181-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86259
O-Subject: [RHEL7 PATCH v2 180/221] KVM: x86: Fix page-tables reserved bits
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Nadav Amit <namit@cs.technion.ac.il>

KVM does not handle the reserved bits of x86 page tables correctly:
In PAE, bits 5:8 are reserved in the PDPTE.
In IA-32e, bit 8 is not reserved.

Signed-off-by: Nadav Amit <namit@cs.technion.ac.il>
Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
(cherry picked from commit cd9ae5fe47dfb9820976c3c38c70f4b07a5a1c36)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/mmu.c b/arch/x86/kvm/mmu.c
index d3aa064..0e6b195 100644
--- a/arch/x86/kvm/mmu.c
+++ b/arch/x86/kvm/mmu.c
@@ -3538,7 +3538,7 @@ static void reset_rsvds_bits_mask(struct kvm_vcpu *vcpu,
  case PT32E_ROOT_LEVEL:
   context->rsvd_bits_mask[0][2] =
    rsvd_bits(maxphyaddr, 63) |
-   rsvd_bits(7, 8) | rsvd_bits(1, 2); /* PDPTE */
+   rsvd_bits(5, 8) | rsvd_bits(1, 2); /* PDPTE */
   context->rsvd_bits_mask[0][1] = exb_bit_rsvd |
    rsvd_bits(maxphyaddr, 62); /* PDE */
   context->rsvd_bits_mask[0][0] = exb_bit_rsvd |
@@ -3550,9 +3550,9 @@ static void reset_rsvds_bits_mask(struct kvm_vcpu *vcpu,
   break;
  case PT64_ROOT_LEVEL:
   context->rsvd_bits_mask[0][3] = exb_bit_rsvd |
-   rsvd_bits(maxphyaddr, 51) | rsvd_bits(7, 8);
+   rsvd_bits(maxphyaddr, 51) | rsvd_bits(7, 7);
   context->rsvd_bits_mask[0][2] = exb_bit_rsvd |
-   rsvd_bits(maxphyaddr, 51) | rsvd_bits(7, 8);
+   rsvd_bits(maxphyaddr, 51) | rsvd_bits(7, 7);
   context->rsvd_bits_mask[0][1] = exb_bit_rsvd |
    rsvd_bits(maxphyaddr, 51);
   context->rsvd_bits_mask[0][0] = exb_bit_rsvd |
-- 
1.7.1